### DISCUSSION TITLE
Typst package ecosystem questions

### DISCUSSION THREAD

**Discussion by neunenak** [typst/typst#232] [General]
I'm currently working on a library for typsetting Leizpig glosses for linguistics papers (https://typst.app/project/rAWHZ7m8MqMRxsFhsVIEDr), and once it is in a somewhat more developed state of development, I'd like to release it as free software code anyone can make use of. This is making me think about concerns pertaining to some kind of Typst package ecosystem that will hopefully develop:

* what sorts of tooling should/will exist for importing external typst packages into your own document? Probably for now library-users will have to manually download `.typ` files they want to use and locally import them into their own projects, but are there plans for some kind of package-management system? What about package versioning?

* what sort of documentation conventions should/will exist around typst packages? Are there plans for Rust-style inline documentation that some tool will convert to documentation (or plans to deliberately avoid this)? 

* I'd like to be able to create a typeset document in typst that outputs some of its own source code as a code block in the final typeset output, without having to copy-paste the same code once where it's actually used and once within a `typst` codeblock. This is probably only useful for creating documents that are in some sense about typst itself, but for that purpose it would definitely come in handy. I couldn't figure out a way to make this happen with typst as it currently stands, and I'm not sure if there is a way to do it at all currently.

**Comment by laurmaedje** [comment]
1. there will be some version of the import syntax where the string designates a package identifier and version
2. something like rustdoc would be nice
3. could you add a show rule for code blocks with tag `typ` that eval's the code and displays it too?

**Reply by neunenak** [reply]
Re: 3, I tried to write a rule that would do this, something like:

````
#show raw.where(lang: "typst-evaled"): it => {
    let txt = it.text
    eval(txt)
    linebreak()
    raw(it.text, lang: "typst")
}

= Leipzig Glossing Examples

This is the classic example of the inflected Georgian verb with an 8-segment consonant cluster:

```typst-evaled
gloss(
    source_text: ([გვ-ფრცქვნ-ი],),
    source_text_style: none,
    transliteration: ([gv-prtskvn-i],),
    morphemes: ([1pl.#obj\-peel-#fmnt],),
    translation: "You peeled us",
)
```
````

which results in the error:
```
error: unknown variable
  ┌─ leipzig-gloss-examples.typ:6:9
  │
6 │     eval(txt)
  │          ^^^
```

**Reply by Heinenen** [reply]
```
#show raw.where(lang: "typst-evaled"): it => {
    let txt = it.text
    eval(repr(txt))
    linebreak()
    raw(it.text, lang: "typc")
}
```
This compiles for me. Also notice that I changed the "lang" to "typc" (for Typst code) 
> Apart from typical language tags known from Markdown, this supports the "typ" and "typc" tags for Typst markup and Typst code, respectively.

@laurmaedje This looks like a bug to me, as other functions (like "type" and "repr") don't complain. 
And both `it.text` and `repr(it.text)` are of type `string`, so it isn't a type-mismatch either.

**Reply by laurmaedje** [reply]
The unknown variable is within the evaluated string.

**Reply by laurmaedje** [reply]
`#obj` and `#fmnt`

**Reply by laurmaedje** [reply]
Note that evaluated strings do not have access to variables from outside the string.

**Reply by Heinenen** [reply]
That makes sense, I misinterpreted the compiler error then.
For completeness, this should work then:

    #show raw.where(lang: "typst-evaled"): it => {
      let txt = it.text
      eval(txt)
      linebreak()
      raw(it.text, lang: "typc")
    }

    = Leipzig Glossing Examples

    This is the classic example of the inflected Georgian verb with an 8-segment consonant cluster:

    ```typst-evaled
      lorem(10)
    ```
